akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala

```diff
@@ -391,9 +391,8 @@ private[projection] class DynamoDBOffsetStore(
           records
             .groupBy(_.pid)
             .valuesIterator
-            .collect {
-              case recordsByPid if !oldState.isDuplicate(recordsByPid.last) => recordsByPid.last
-            }
+            .map(_.maxBy(_.seqNr))
+            .filterNot(oldState.isDuplicate _)
```

Comment:
This seems to assume that the latest record by pid would be the last in the group, but I don't see callers actually ensuring that (e.g. if a particular batch of records includes the same pid from forward querying and then backtracking).

Reply:
The records for a pid should be in order, so this should work as is. Same logic in r2dbc.